Enable shared OSX tests

MacOS 10.15.7, Xcode-12.4 (probably not relevant here, but...).

Initially, I could not run tests because the make-script could not load dynamic library. *Problem with the upstream, unresolved.*

Changing the build to producing static library, the tests can run - but some are skipped with this message:
```
test_tlsextms needs the dynamic engine feature enabled
```

What's this "dynamic engine feature"? How do I enable it - just by removing `no-shared` from the `./Configuration ...` command?

And why the README tells to build this fork with `no-shared`? Is this restriction still valid?

BTW (this probably relates to the previous issue - I'll move it elsewhere if you request), can you explain this:
```
$ env | grep DYLD_LIBRARY_PATH
DYLD_LIBRARY_PATH=/Users/ur20980/src/liboqs/build/lib:/Users/ur20980/src/oqs-openssl:/opt/X11/lib:/opt/local/lib:/usr/local/lib:/usr/lib:/Users/ur20980/lib::/Library/Java/NSS/lib::.
$ make test VFO=1 'TESTS=test-tls13ccs'DYLD_LIBRARY_PATH=".:${HOME}/oqs-openssl/lib:/opt/local/lib:${DYLD_LIBRARY_PATH}" make test VFO=1 'TESTS=tls13ccs'
make depend && make _tests
make[1]: Entering directory '/Users/ur20980/src/oqs-openssl'
make[1]: Leaving directory '/Users/ur20980/src/oqs-openssl'
make[1]: Entering directory '/Users/ur20980/src/oqs-openssl'
DEBUG[Makefile] DYLD_LIBRARY_PATH=
( cd test; \
  mkdir -p test-runs; \
  SRCTOP=../. \
  BLDTOP=../. \
  RESULT_D=test-runs \
  PERL="/opt/local/bin/perl5.26" \
  EXE_EXT= \
  OPENSSL_ENGINES=`cd .././engines 2>/dev/null && pwd` \
  OPENSSL_DEBUG_MEMORY=on \
    /opt/local/bin/perl5.26 .././test/run_tests.pl tls13ccs )
Test tls13ccs found no match, skipping addition...
Use of uninitialized value $ENV{"DYLD_LIBRARY_PATH"} in concatenation (.) or string at .././test/run_tests.pl line 86.
DEBUG[run_tests.pl] DYLD_LIBRARY_PATH=
Files=0, Tests=0,  0 wallclock secs ( 0.00 usr +  0.00 sys =  0.00 CPU)
Result: NOTESTS
make[1]: Leaving directory '/Users/ur20980/src/oqs-openssl'
make: *** No rule to make target 'make'.  Stop.
```

Clearly, `DYLD_LIBRARY_PATH` env var is set, and equally clearly, perl script invoked by `make test` does not see it. 

???????

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -31,6 +31,10 @@ jobs:
         description: "Build OpenSSL as a shared library."
         type: boolean
         default: false
+      TESTSET:
+        description: "OQS test set."
+        type: string
+        default: oqs-test/test_tls_basic.py
     docker:
       - image: openquantumsafe/ci-ubuntu-focal-x86_64:latest
 # Re-enable iff docker enforces rate limitations without auth:
@@ -52,8 +56,8 @@ jobs:
           name: Run OpenSSL tests
           command: env LD_LIBRARY_PATH="$(pwd):$(pwd)/oqs/lib" make test
       - run:
-          name: Run OQS connection, CMS and speed tests
-          command: env LD_LIBRARY_PATH="$(pwd):$(pwd)/oqs/lib" python3 -m pytest --numprocesses=auto oqs-test/test_tls_full.py oqs-test/test_cms.py oqs-test/test_speed.py
+          name: Run OQS tests
+          command: env LD_LIBRARY_PATH="$(pwd):$(pwd)/oqs/lib" python3 -m pytest --numprocesses=auto << parameters.TESTSET >>
 
   macOS:
     description: A template for running OQS-OpenSSL tests on macOS
@@ -68,6 +72,10 @@ jobs:
         description: "Build OpenSSL as a shared library."
         type: boolean
         default: false
+      TESTSET:
+        description: "OQS test set."
+        type: string
+        default: oqs-test/test_tls_basic.py
     steps:
       - checkout # change this from "checkout" to "*localCheckout" when running CircleCI locally
       - run:
@@ -87,8 +95,8 @@ jobs:
           name: Run OpenSSL tests
           command: ln -sf oqs/lib/liboqs.0.dylib . && env DYLD_LIBRARY_PATH="$(pwd):$(pwd)/oqs/lib" make test
       - run:
-          name: Run OQS connection, CMS and speed tests
-          command: env DYLD_LIBRARY_PATH="$(pwd):$(pwd)/oqs/lib" python3 -m pytest --numprocesses=auto oqs-test/test_tls_full.py oqs-test/test_cms.py oqs-test/test_speed.py
+          name: Run OQS tests
+          command: env DYLD_LIBRARY_PATH="$(pwd):$(pwd)/oqs/lib" python3 -m pytest --numprocesses=auto << parameters.TESTSET >>
 
   ubuntu_boringssl_interop:
     description: A job that tests interoperability between OQS-OpenSSL and OQS-BoringSSL on a x64 Ubuntu Bionic Docker VM
@@ -152,6 +160,10 @@ workflows:
       - ubuntu_focal:
           name: ubuntu-static_oqs-static_ossl
           context: openquantumsafe
+      - macOS:
+          name: macOS-shared_oqs-shared_ossl
+          LIBOQS_SHARED: true
+          OPENSSL_SHARED: true
       - testapproval:
           requires:
             - ubuntu-static_oqs-static_ossl
@@ -159,15 +171,13 @@ workflows:
       - macOS:
           <<: *require_testapproval
           name: macOS-static_oqs-static_ossl
-      - macOS:
-          name: macOS-shared_oqs-shared_ossl
-          LIBOQS_SHARED: true
-          OPENSSL_SHARED: true
+          TESTSET: oqs-test/test_tls_full.py oqs-test/test_cms.py oqs-test/test_speed.py
       - ubuntu_focal:
           name: ubuntu-shared_oqs-shared_ossl
           context: openquantumsafe
           LIBOQS_SHARED: true
           OPENSSL_SHARED: true
+          TESTSET: oqs-test/test_tls_full.py oqs-test/test_cms.py oqs-test/test_speed.py
       - ubuntu_boringssl_interop:
           <<: *require_testapproval
           name: boringssl-interop

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -173,6 +173,7 @@ workflows:
           name: macOS-static_oqs-static_ossl
           TESTSET: oqs-test/test_tls_full.py oqs-test/test_cms.py oqs-test/test_speed.py
       - ubuntu_focal:
+          <<: *require_testapproval
           name: ubuntu-shared_oqs-shared_ossl
           context: openquantumsafe
           LIBOQS_SHARED: true

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -54,7 +54,7 @@ jobs:
   macOS:
     description: A template for running OQS-OpenSSL tests on macOS
     macos:
-      xcode: "11.3.0"
+      xcode: "12.4.0"
     parameters:
       LIBOQS_SHARED:
         description: "Build liboqs as a shared library."
@@ -78,11 +78,10 @@ jobs:
       - run:
           name: Build OpenSSL
           command: <<# parameters.OPENSSL_SHARED >> env OPENSSL_LIBTYPE=shared <</ parameters.OPENSSL_SHARED >> ./oqs-scripts/build_openssl.sh
-      # TODO: Figure out why the tests fail on macOS when OpenSSL
-      # is built as a shared library.
-      #- run:
-          #name: Run OpenSSL tests
-          #command: env DYLD_LIBRARY_PATH="$(pwd):$(pwd)/oqs/lib" make test
+      # making liboqs shared lib available in a non-standard place due to https://github.com/openssl/openssl/issues/13413 to allow shared-lib make test to run 
+      - run:
+          name: Run OpenSSL tests
+          command: ln -sf oqs/lib/liboqs.0.dylib . && env DYLD_LIBRARY_PATH="$(pwd):$(pwd)/oqs/lib" make test
       - run:
           name: Run OQS connection, CMS and speed tests
           command: env DYLD_LIBRARY_PATH="$(pwd):$(pwd)/oqs/lib" python3 -m pytest --numprocesses=auto oqs-test/test_tls_full.py oqs-test/test_cms.py oqs-test/test_speed.py

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -1,5 +1,9 @@
 version: 2.1
 
+require_testapproval: &require_testapproval
+  requires:
+    - testapproval
+
 parameters:
   run_downstream_tests:
     type: boolean
@@ -16,7 +20,7 @@ localCheckout: &localCheckout
     cp -a /tmp/_circleci_local_build_repo/.git ${PROJECT_PATH}
 
 jobs:
-  ubuntu_bionic:
+  ubuntu_focal:
     description: A template for running OQS-OpenSSL tests on x64 Ubuntu Bionic Docker VMs
     parameters:
       LIBOQS_SHARED:
@@ -28,7 +32,7 @@ jobs:
         type: boolean
         default: false
     docker:
-      - image: openquantumsafe/ci-ubuntu-bionic-x86_64:latest
+      - image: openquantumsafe/ci-ubuntu-focal-x86_64:latest
 # Re-enable iff docker enforces rate limitations without auth:
 #        auth:
 #          username: $DOCKER_LOGIN
@@ -89,7 +93,7 @@ jobs:
   ubuntu_boringssl_interop:
     description: A job that tests interoperability between OQS-OpenSSL and OQS-BoringSSL on a x64 Ubuntu Bionic Docker VM
     docker:
-      - image: openquantumsafe/ci-ubuntu-bionic-x86_64:latest
+      - image: openquantumsafe/ci-ubuntu-focal-x86_64:latest
 # Re-enable iff docker enforces rate limitations without auth:
 #        auth:
 #          username: $DOCKER_LOGIN
@@ -145,21 +149,27 @@ workflows:
   version: 2.1
   build:
     jobs:
+      - ubuntu_focal:
+          name: ubuntu-static_oqs-static_ossl
+          context: openquantumsafe
+      - testapproval:
+          requires:
+            - ubuntu-static_oqs-static_ossl
+          type: approval
       - macOS:
+          <<: *require_testapproval
           name: macOS-static_oqs-static_ossl
       - macOS:
           name: macOS-shared_oqs-shared_ossl
           LIBOQS_SHARED: true
           OPENSSL_SHARED: true
-      - ubuntu_bionic:
-          name: ubuntu-static_oqs-static_ossl
-          context: openquantumsafe
-      - ubuntu_bionic:
+      - ubuntu_focal:
           name: ubuntu-shared_oqs-shared_ossl
           context: openquantumsafe
           LIBOQS_SHARED: true
           OPENSSL_SHARED: true
       - ubuntu_boringssl_interop:
+          <<: *require_testapproval
           name: boringssl-interop
           context: openquantumsafe
   on-main-branch:

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -68,7 +68,7 @@ jobs:
       - checkout # change this from "checkout" to "*localCheckout" when running CircleCI locally
       - run:
           name: Install dependencies
-          command: brew unlink python@2 && env HOMEBREW_NO_AUTO_UPDATE=1 brew install cmake ninja && pip3 install pytest pytest-xdist psutil
+          command: env HOMEBREW_NO_AUTO_UPDATE=1 brew install cmake ninja && pip3 install pytest pytest-xdist psutil
       - run:
           name: Clone liboqs
           command: ./oqs-scripts/clone_liboqs.sh

